src/main/scala/xiangshan/frontend/PrunedAddr.scala

```diff
@@ -0,0 +1,72 @@
+/***************************************************************************************
+ * Copyright (c) 2020-2021 Institute of Computing Technology, Chinese Academy of Sciences
+ * Copyright (c) 2020-2021 Peng Cheng Laboratory
+ *
+ * XiangShan is licensed under Mulan PSL v2.
+ * You can use this software according to the terms and conditions of the Mulan PSL v2.
+ * You may obtain a copy of Mulan PSL v2 at:
+ *          http://license.coscl.org.cn/MulanPSL2
+ *
+ * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
+ * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
+ *
+ * See the Mulan PSL v2 for more details.
+ ***************************************************************************************/
+
+package xiangshan.frontend
+
+import chisel3._
+import chisel3.util._
+import org.chipsalliance.cde.config.Parameters
+import xiangshan.XSBundle
+
+class PrunedAddr(val length: Int)(implicit p: Parameters) extends XSBundle {
+  val addr: UInt = UInt((length - instOffsetBits).W)
```

Comment:
> According to RISC-V specification, the least significant 1 or 2 bits (depending on whether the C extension is enabled or not) are hardwired to zero. Therefore, these bits can be pruned to save area.

That's true for xepc:
> The low bit of mepc (mepc[0]) is always zero. On implementations that support only IALIGN=32, the two low bits (mepc[1:0]) are always zero.

But NOT for jal, jalr and branches:
> The jump and link (JAL) instruction uses the J-type format, where the J-immediate encodes a signed offset in multiples of 2 bytes. 
> ...
> The indirect jump instruction JALR (jump and link register) uses the I-type encoding. The target address is obtained by adding the sign-extended 12-bit I-immediate to the register rs1, then setting the least-significant bit of the result to zero.
> ...
> The JAL and JALR instructions will generate an instruction-address-misaligned exception if the target address is not aligned to a four-byte boundary.
> > Instruction-address-misaligned exceptions are not possible on machines that support extensions with 16-bit aligned instructions, such as the compressed instruction-set extension, C.
> 
> ...
> All branch instructions use the B-type instruction format. The 12-bit B-immediate encodes signed offsets in multiples of 2 bytes.

So I guess this should be `UInt((length - 1).W)`.

Btw, `!HasCExtension` support seems to be broken now, we don't have "Instruction address misaligned" check, should we try to fix that (maybe later in another PR)? What do you think @eastonman @Yan-Muzi

Reply:
That's true, but misaligned exceptions should be generated in backend, so it is safe for frontend to prune `instOffsetBits` bits in the address.

Reply:
> That's true, but misaligned exceptions should be raised in backend, so it is safe for frontend to prune `instOffsetBits` bits in the address.

Maybe they should check (I think it makes more sense to check that in frontend), but AFAIK the current JumpUnit implementation doesn't.

Also, even if they did check it, this would require frontend to have the appropriate datapath support (since misaligned should be marked on the jump target, not jal/jalr/branch itself), as is the case now with `backendIAf` etc., which the frontend doesn't have right now.

Reply:
> > That's true, but misaligned exceptions should be raised in backend, so it is safe for frontend to prune `instOffsetBits` bits in the address.
> 
> Maybe they should check (I think it makes more sense to check that in frontend), but AFAIK the current JumpUnit implementation doesn't.
> 
> Also, even if they did check it, this would require frontend to have the appropriate datapath support (since misaligned should be marked on the jump target, not jal/jalr/branch itself), as is the case now with `backendIAf` etc., which the frontend doesn't have right now.

Re-reading the manual, you're right that misaligned should be marked on jal/jalr/branch itself, but strangely the current backend implementation really doesn't do this check.